### PR TITLE
add stored-atom

### PR DESCRIPTION
**Pull request by adrianmroz**

### CODE REVIEW

File: src/lensed-atom.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { IAtom, Atom } from './atom';
+import { Modifier } from './modifier';
+import { Projection } from './projection';
+import { Observable } from 'rxjs';
+import { over, view, compose } from 'ramda';
+import { Lens } from '@types/ramda';
+
+export class LensedAtom<T, S> implements IAtom<T> {
+  constructor(private lens: Lens, private source: Atom<S>) {
+  }
```
Comment (bigsergey) [inline src/lensed-atom.ts]: empty constructor? is it okey?

Comment (erykpiast) [reply src/lensed-atom.ts]: It's not empty actually. `(private lens: Lens, private source: Atom<S>)` part sets two private properties on class instance (named as arguments: `lens` and `source`). TypeScript magic ;)

File: src/atom.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import { ConnectableObservable, Observable, Subject, Subscription } from 'rxjs';
+import { Lens } from '@types/ramda';
+import { Modifier } from './modifier';
+import { Projection } from './projection';
+import { LensedAtom } from './lensed-atom';
+
+export interface IAtom<T> {
```
Comment (erykpiast) [inline src/atom.ts]: I would say, this declaration could be moved to separate file along with `Modifier` and `Projection` types, as those two may feel a bit lonely.

File: src/stored-atom.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  from?: (source: any) => T,
+  to?: (value: T) => any
+}
+
+export class StoredAtom<T> extends Atom<T> {
+  private storedValue$: Observable<T>;
+
+  constructor(defaultValue: T, {key, from = identity, to = identity}: StoredAtomOptions<T>) {
+    const savedItem = localStorage.getItem(key);
+    super(savedItem === null ? defaultValue : from(JSON.parse(savedItem)));
```
Comment (erykpiast) [inline src/stored-atom.ts]: `savedItem === null ? defaultValue : from(JSON.parse(savedItem))` name it

Comment (erykpiast) [reply src/stored-atom.ts]: I'd say this one: `from(JSON.parse(savedItem))` should be called `parse` and this `SON.stringify(to(value))` - `serialize`. They may be just functions in constructor scope, but named ones. If you do that expression like `savedItem === null ? defaultValue : parse(savedItem)` is clear enough to have no name.

File: index.ts [hunk 4]
```diff
@@ -1,82 +1,7 @@
-import { Observable, Subject, ConnectableObservable, Subscription } from 'rxjs';
-import { over, view, compose } from 'ramda';
-import { Lens} from '@types/ramda';
+import { Atom as AtomDef } from './src/atom';
```
Comment (erykpiast) [inline index.ts]: `export { Atom } from './src/atom'?`

File: src/atom.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  get(): Observable<T>;
+  modify(fn: Modifier<T>): void;
+  project<R>(projection: Projection<T, R>): Observable<R>;
+  view<S, A>(lens: Lens): LensedAtom<S, A>;
+}
+
+export class Atom<T> implements IAtom<T> {
```
Comment (erykpiast) [inline src/atom.ts]: shouldn't it be default export?

File: src/lensed-atom.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { IAtom, Atom } from './atom';
+import { Modifier } from './modifier';
+import { Projection } from './projection';
+import { Observable } from 'rxjs';
+import { over, view, compose } from 'ramda';
+import { Lens } from '@types/ramda';
+
+export class LensedAtom<T, S> implements IAtom<T> {
```
Comment (erykpiast) [inline src/lensed-atom.ts]: default export?

File: src/stored-atom.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import { identity } from 'ramda';
+
+interface StoredAtomOptions<T> {
+  key: string;
+  from?: (source: any) => T,
+  to?: (value: T) => any
+}
+
+export class StoredAtom<T> extends Atom<T> {
+  private storedValue$: Observable<T>;
```
Comment (erykpiast) [inline src/stored-atom.ts]: default export?

File: src/stored-atom.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { Atom } from './atom';
+import { Observable } from 'rxjs';
+import { identity } from 'ramda';
+
+interface StoredAtomOptions<T> {
```
Comment (erykpiast) [inline src/stored-atom.ts]: If you started with `IName` convention for interfaces, let's stick to it or abandon at all.

File: src/modifier.ts [hunk 9]
```diff
@@ -0,0 +1 @@
+export type Modifier<T> = (value: T) => T;
```
Comment (erykpiast) [inline src/modifier.ts]: default export (if you don't gonna merge it with `projection.ts`)

File: src/projection.ts [hunk 10]
```diff
@@ -0,0 +1 @@
+export type Projection<T, R> = (source: T) => R;
```
Comment (erykpiast) [inline src/projection.ts]: as above

**Review by erykpiast — CHANGES_REQUESTED** [review]
How could you refactor without any tests? :P

**Pull request merged**